### DISCUSSION TITLE
[feature request] relation column filter

### DISCUSSION THREAD

**Discussion by gpedro** [Power-Components/livewire-powergrid#1086] [Ideas]
# ⚡ PowerGrid ⚡ Feature Request

## Summary

I made a little helper to make filtering the relationship columns easier and that maybe should be included in the core package.

### Suggested Solution(s)

Usage:
```php
Filter::inputText('category_name')->builder($this->builderRelation('categories', 'name')),
```
Helper:
```php

    private function builderRelation($relation, $field): Closure
    {
        return function ($query, $opts) use ($relation, $field) {
            $operator = '=';
            $value = $opts['value'];
            $raw = false;

            switch ($opts['selected']) {
                case 'is':
                    $operator = '=';
                    $value = $opts['value'];
                    break;

                case 'is_not':
                    $operator = '!=';
                    $value = $opts['value'];
                    break;

                case 'starts_with':
                    $operator = SqlSupport::like($query);
                    $value = $opts['value'] . '%';
                    break;

                case 'ends_with':
                    $operator = SqlSupport::like($query);
                    $value = '%' .$opts['value'];
                    break;

                case 'contains':
                    $operator = SqlSupport::like($query);
                    $value = '%' . $opts['value']. '%';
                    break;

                case 'contains_not':
                    $operator =  'NOT ' . SqlSupport::like($query);
                    $value = '%' . $opts['value']. '%';
                    break;

                case 'is_empty':
                    $raw = true;
                    $operator = '=';
                    $value = '';
                    $query->whereRelationIsNull($relation, $field);
                    break;

                case 'is_not_empty':
                    $operator = '!=';
                    $value = '';
                    $query->whereRelationIsNotNull($relation, $field);
                    break;

                case 'is_null':
                    $raw = true;
                    $query->whereRelationIsNull($relation, $field);
                    break;

                case 'is_null':
                    $raw = true;
                    $query->whereRelationIsNotNull($relation, $field);
                    break;

                case 'is_blank':
                    $operator = '=';
                    $value = '';
                break;

                case 'is_not_blank':
                    $operator = '1=';
                    $value = '';
                break;
            };

            if ($raw) {
                return;
            }

            $query->whereRelation($relation, $field, $operator, $value);
        };
    }
```

**Comment by CrazyBoy49z** [comment]
#359

**Comment by luanfreitasdev** [comment]
> # ⚡ PowerGrid ⚡ Feature Request
> ## Summary
> I made a little helper to make filtering the relationship columns easier and that maybe should be included in the core package.
> 
> ### Suggested Solution(s)
> Usage:
> 
> ```
> Filter::inputText('category_name')->builder($this->builderRelation('categories', 'name')),
> ```
> 
> Helper:
> 
> ```
>     private function builderRelation($relation, $field): Closure
>     {
>         return function ($query, $opts) use ($relation, $field) {
>             $operator = '=';
>             $value = $opts['value'];
>             $raw = false;
> 
>             switch ($opts['selected']) {
>                 case 'is':
>                     $operator = '=';
>                     $value = $opts['value'];
>                     break;
> 
>                 case 'is_not':
>                     $operator = '!=';
>                     $value = $opts['value'];
>                     break;
> 
>                 case 'starts_with':
>                     $operator = SqlSupport::like($query);
>                     $value = $opts['value'] . '%';
>                     break;
> 
>                 case 'ends_with':
>                     $operator = SqlSupport::like($query);
>                     $value = '%' .$opts['value'];
>                     break;
> 
>                 case 'contains':
>                     $operator = SqlSupport::like($query);
>                     $value = '%' . $opts['value']. '%';
>                     break;
> 
>                 case 'contains_not':
>                     $operator =  'NOT ' . SqlSupport::like($query);
>                     $value = '%' . $opts['value']. '%';
>                     break;
> 
>                 case 'is_empty':
>                     $raw = true;
>                     $operator = '=';
>                     $value = '';
>                     $query->whereRelationIsNull($relation, $field);
>                     break;
> 
>                 case 'is_not_empty':
>                     $operator = '!=';
>                     $value = '';
>                     $query->whereRelationIsNotNull($relation, $field);
>                     break;
> 
>                 case 'is_null':
>                     $raw = true;
>                     $query->whereRelationIsNull($relation, $field);
>                     break;
> 
>                 case 'is_null':
>                     $raw = true;
>                     $query->whereRelationIsNotNull($relation, $field);
>                     break;
> 
>                 case 'is_blank':
>                     $operator = '=';
>                     $value = '';
>                 break;
> 
>                 case 'is_not_blank':
>                     $operator = '1=';
>                     $value = '';
>                 break;
>             };
> 
>             if ($raw) {
>                 return;
>             }
> 
>             $query->whereRelation($relation, $field, $operator, $value);
>         };
>     }
> ```

Added https://github.com/Power-Components/livewire-powergrid/commit/2778bd113305bf4099c6cf48e37c17c4d23d2194
